aten/src/ATen/CPUGenerator.cpp

```diff
@@ -1,49 +1,88 @@
 #include <ATen/CPUGenerator.h>
-
-#define const_generator_cast(generator) \
-  dynamic_cast<const CPUGenerator&>(generator)
+#include <c10/util/C++17.h>
 
 namespace at {
 
-CPUGenerator::CPUGenerator(Context * context_)
-  : context(context_), generator(THGenerator_new())
-{}
+namespace detail {
 
-CPUGenerator::~CPUGenerator() {
-  if (generator)
-    THGenerator_free(generator);
+/*
+* call once pattern to initialize default generator once
+*/
+CPUGenerator& getDefaultCPUGenerator() {
+  std::call_once(cpu_device_flag, [&] {
+    default_gen_cpu = c10::guts::make_unique<CPUGenerator>(default_rng_seed_val);
+  });
+  return *default_gen_cpu;
 }
 
-CPUGenerator& CPUGenerator::copy(const Generator& from) {
-  THGenerator_copy(generator, const_generator_cast(from).generator);
-  return *this;
+/*
+* Utility to create a CPUGenerator. Returns a unique_ptr
+*/
+std::unique_ptr<CPUGenerator> createCPUGenerator(uint64_t seed_val) {
+  return c10::guts::make_unique<CPUGenerator>(seed_val);
 }
 
-CPUGenerator& CPUGenerator::free() {
-  THGenerator_free(generator);
-  return *this;
-}
+} //namespace detail
 
-uint64_t CPUGenerator::seed() {
-  return THRandom_seed(generator);
-}
+/*
+* CPUGenerator class implementation
+*/
+CPUGenerator::CPUGenerator(uint64_t seed_in)
+  : Generator(Device(DeviceType::CPU), seed_in), engine_(Philox4_32_10(seed_in)) {}
 
-uint64_t CPUGenerator::initialSeed() {
-  return THRandom_initialSeed(generator);
-}
+CPUGenerator::CPUGenerator(const CPUGenerator& other)
+  : CPUGenerator(other, 
+                 std::lock_guard<std::mutex>(other.mutex)) {}
 
-CPUGenerator& CPUGenerator::manualSeed(uint64_t seed) {
-  THRandom_manualSeed(generator, seed);
+CPUGenerator::CPUGenerator(const CPUGenerator &other, 
+                           const std::lock_guard<std::mutex> &other_mutex)
+  : Generator(other, other_mutex), engine_(other.engine_) {}
+
+CPUGenerator::CPUGenerator(CPUGenerator&& other)
+  : CPUGenerator(other,
+                 std::lock_guard<std::mutex>(other.mutex)) {}
+
+CPUGenerator::CPUGenerator(const CPUGenerator &&other,
+                           const std::lock_guard<std::mutex> &other_mutex)
```

Comment:
We should resolve the question of thread-safety first (though I agree that copies should be mutex protected), but I do wonder about how much this idiom of taking `lock_guard` as an argument is giving us. Is this a recommendation somewhere? :)

Reply:
I was just paranoid about thread safety and just followed this: https://www.justsoftwaresolutions.co.uk/threading/thread-safe-copy-constructors.html. Also just wanted copy constructors but having the mutex wouldn't give me auto generated ones, so did the whole google search how to get my copy constructors thread safe.

Reply:
I guess this is moot since we're killing all the copy constructors